docs/decisions/0040-OTel-semantic-convention.md

```diff
@@ -0,0 +1,460 @@
+    // For streaming, collect a list of `StreamingTextContent` and concatenate them into a single `TextContent` at the end of the streaming.
+    void SetCompletionResponses(Activity? activity, IReadOnlyList<TextContent> completions, CompletionUsage? usage, PromptExecutionSettings? executionSettings)
+    {
+        ...
+    }
+
+    // Can be used for both non-streaming endpoints and streaming endpoints.
+    // For streaming, collect a list of `StreamingChatMessageContent` and concatenate them into a single `ChatMessageContent` at the end of the streaming.
+    void SetCompletionResponses(Activity? activity, IReadOnlyList<ChatMessageContent> completions, CompletionUsage? usage, PromptExecutionSettings? executionSettings)
+    {
+        ...
+    }
+}
+```
+
+Example usage
+
+```C#
+// In client
+public async Task<IReadOnlyList<TextContent>> GenerateTextAsync(
+    string prompt,
+    PromptExecutionSettings? executionSettings,
+    CancellationToken cancellationToken)
+{
+    using var activity = ModelDiagnostics.StartCompletionActivity(
+        $"HuggingFace {this._modelId}",
+        this._modelId,
+        "HuggingFace",
+        prompt,
+        executionSettings);
+
+    var completions = ...;
+    // Usage can be estimated.
+    var usage = ...;
+
+    ModelDiagnostics.SetCompletionResponses(
+        activity,
+        completions,
+        usage,
+        executionSettings);
+
+    return completions;
+}
+
+// In service
+public Task<IReadOnlyList<TextContent>> GetTextContentsAsync(string prompt, PromptExecutionSettings? executionSettings = null, Kernel? kernel = null, CancellationToken cancellationToken = default)
+{
+    using var activity = ModelDiagnostics.StartCompletionActivity(
+        $"HuggingFace {this.AttributesInternal[AIServiceExtensions.ModelIdKey]}",
+        this.AttributesInternal[AIServiceExtensions.ModelIdKey],
+        "HuggingFace",
+        prompt,
+        executionSettings);
+
+    var completions = ...;
+
+    // Usage is usually not returned from the clients.
+    ModelDiagnostics.SetCompletionResponses(
+        activity,
+        completions,
+        null,
+        executionSettings);
+
+    return completions;
+}
+```
+
+#### Dependency injection
+
+The interface and the implementation will live under `SemanticKernel.Abstractions/AI`.
+
+`DefaultModelInstrumentation : IModelInstrumentation`
+
+```C#
+public interface IModelInstrumentation
+{
+    Activity? StartCompletionActivity(string name, string modelName, string modelProvider, ChatHistory chatHistory, PromptExecutionSettings? executionSettings);
+
+    Activity? StartCompletionActivity(string name, string modelName, string modelProvider, string prompt, PromptExecutionSettings? executionSettings);
+
+    void SetCompletionResponses(Activity? activity, IReadOnlyList<TextContent> completions, CompletionUsage? usage);
+
+    void SetCompletionResponses(Activity activity, IReadOnlyList<ChatMessageContent> completions, CompletionUsage? usage);
+}
```

Comment:
Why does any of this need to be exposed right now?

What would another implementation of IModelInstrumentation do differently from the default that it needs to be configurable?

Are you viewing `builder.Services.AddSingleton<IModelInstrumentation, DefaultModelInstrumentation>();` as the way of opt-ing in to this information being included?

Reply:
A developer can attach more tags/attributes depending on his or her use cases in a custom implementation. 

And yes, I see `builder.Services.AddSingleton<IModelInstrumentation, DefaultModelInstrumentation>();` as the way to include LLM request activities in the traces.

Reply:
> A developer can attach more tags/attributes depending on his or her use cases in a custom implementation.

And to do that they then need to provide a full implementation of this interface? How does it compose, e.g. if I want to use both the default implementation but then I also want to add this or that tag?

Reply:
Yeah, that's a good use case. Right now, they'd need to rewrite everything. I will need to think about if this flexibility adds value, and if so how to go about it.

Reply:
I actually think we should not let users opt in to have this information included in the traces. It should be included by default. And the way it works in .Net is that if [ActivitySource.StartActivity](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.activitysource.startactivity) internally determines if there are any listeners recording the Activity. If there are no registered listeners or there are listeners that are not interested, no activities will be created.

Another option is to create static ModelDiagnostics member in each service/client. This approach brings the following benefits:
1. It still follows the [best practice](https://learn.microsoft.com/en-us/dotnet/core/diagnostics/distributed-tracing-instrumentation-walkthroughs#best-practices-1) to create an activity source once within a component.
2. It makes it possible to have multiple activity sources with different names, e.g. `Microsoft.SemanticKernel.Connectors.HuggingFace` and `Microsoft.SemanticKernel.Connectors.Google`.

I will also include this in the ADR.

Reply:
Yes, we should be doing all the main tracing/events/metrics by default. We will want all of the sensitive stuff to be opt-in, though, e.g. including the actual prompt as a tag on an activity.  There are a variety of ways that could be exposed.

As for adding additional information beyond that, I think it'd be better to model it similar to how the other filters are being modeled, basically a callback that gives you the activity and other relevant information (e.g. the Kernel, KernelFunction, KernelArguments, the prompt, whatever), and you can tag the activity however you like. If you have multiple filters, they'd all have an opportunity to get in on it. As with other filters, you could register these in DI, or just add them to the Kernel directly. Opting in to having the sensitive data traced could just be an extra built-in implementation of that interface, such that it'll be invoked like any others.  Or something like that.

Reply:
Updated. 

I agree. According to the semantic conventions, all the sensitive stuff (currently only includes prompts and responses) will be events, which are optional and configurable. I have listed two options in the ADR as to how to configure it (through `IsEventCollectionEnable`). 

For additional information, if we go the filter route and would like to trace information at the function level, I think it's probably worth use separate activities, as it is no longer relevant to the LLM requests. We actually have that implemented already: https://github.com/microsoft/semantic-kernel/blob/main/dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunction.cs#L159.

Reply:
> if we go the filter route

Why would that need separate activities?

Reply:
1. Users who want to view LLM request details may not be interested in seeing kernel function details.
2. Activities that trace LLM requests contain a special set of attributes standardized by the LLM semantic conventions.
3. Even if the above two aren't enough to justify the separation, kernel functions and connectors/models are decoupled entities. Of course, we can attach attributes in connectors to activities created in kernel functions by using `Activity.Current?`. I personally think that makes the code harder to maintain.

Reply:
But that's not filters requiring separate activities. We can choose to use multiple activities, and doing so makes sense, if we want to have one that covers all function invocation and another per service call, but that's orthogonal to augmenting them with additional tags. Filters can be invoked for each activity, and the filter can decide what to add to what.

Reply:
Agree. The interface approach appears to offer minimal added value for users while generating additional confusion, so I have removed it.